lib/integrations/awsoidc/deployservice.go

```diff
@@ -0,0 +1,696 @@
+	if len(clusters) == 0 {
+		return true
+	}
+
+	cluster := clusters[0]
+
+	return *cluster.Status == clusterStatusInactive
+}
+
+// waitForActiveCluster waits until the Cluster is Active.
+// If the Cluster is Provisioning, then it waits at most clusterStatusProvisioningWaitTime (30 seconds) for it to become ready.
+func waitForActiveCluster(ctx context.Context, clt DeployServiceClient, req DeployServiceRequest, cluster *ecsTypes.Cluster) error {
+	if cluster.Status != nil && *cluster.Status == clusterStatusActive {
+		return nil
+	}
+
+	retry, err := retryutils.NewConstant(clusterStatusProvisioningWaitTimeTick)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	retryCtx, cancel := context.WithTimeout(ctx, clusterStatusProvisioningWaitTime)
+	defer cancel()
+
+	err = retry.For(retryCtx, func() error {
+		describeClustersResponse, err := clt.DescribeClusters(ctx, &ecs.DescribeClustersInput{
+			Clusters: []string{*req.ClusterName},
+		})
+		if err != nil {
+			return retryutils.PermanentRetryError(trace.Wrap(err))
+		}
+
+		if len(describeClustersResponse.Clusters) == 0 {
+			return retryutils.PermanentRetryError(trace.NotFound("cluster %q does not exist", *cluster.ClusterName))
+		}
+
+		cluster := describeClustersResponse.Clusters[0]
+		if cluster.Status == nil {
+			return retryutils.PermanentRetryError(trace.Errorf("cluster %q has an unknown (nil) status", *cluster.ClusterName))
+		}
+
+		if *cluster.Status == clusterStatusActive {
+			return nil
+		}
+
+		if *cluster.Status == clusterStatusProvisioning {
+			return trace.Errorf("cluster %q is provisioning...", *cluster.ClusterName)
+		}
+
+		return retryutils.PermanentRetryError(trace.Errorf("unexpected status %s for ECS Cluster %q", *cluster.ClusterName, *cluster.Status))
+	})
+
+	return trace.Wrap(err)
+}
+
+// upsertService creates or updates the service.
+// If the service exists but its LaunchType is not Fargate, then it gets re-created.
+func upsertService(ctx context.Context, clt DeployServiceClient, req DeployServiceRequest, taskARN string) (*ecsTypes.Service, error) {
+	describeServiceOut, err := clt.DescribeServices(ctx, &ecs.DescribeServicesInput{
+		Services: []string{*req.ServiceName},
+		Cluster:  req.ClusterName,
+		Include: []ecsTypes.ServiceField{
+			ecsTypes.ServiceFieldTags,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Service already exists.
+	if len(describeServiceOut.Services) > 0 {
+		service := &describeServiceOut.Services[0]
+
+		if service.Status == nil {
+			return nil, trace.Errorf("unknown status for ECS Service %q", *req.ServiceName)
+		}
+
+		if *service.Status == serviceStatusDraining {
+			return nil, trace.Errorf("ECS Service is draining, please retry in a couple of minutes")
+		}
+
+		if *service.Status == serviceStatusActive {
+			ownershipTags := req.ResourceCreationTags
+			if !ownershipTags.MatchesECSTags(service.Tags) {
+				return nil, trace.Errorf("ECS Service %q already exists but is not managed by Teleport. "+
```

Comment:
Similar question to above, what's the use-case for this?

Reply:
Marking this as resolved, so we can focus the discussion on the conversation above.